File: day9/solution.py
from collections import defaultdict
from functools import reduce
import sys

# Adding walls of 9 around input
# 999999
# 9....9
# 9....9
# 999999
M = 0
N = 0
matrix = []
with open('input') as input:
  for line in input:
    if not matrix:
      N = len(line.strip()) + 2
      matrix.append(['9']*N)
    n = [x for x in line.strip()]
    n.insert(0, '9')
    n.append('9')
    matrix.append(n)
    M += 1
  matrix.append(['9']*N)
  M += 2

positions = set([(i, j) for j in range(1, N-1) for i in range(1, M-1)])

def flood(pos):
  sum = 0
  children = [pos]
  i, j = pos
  matrix[i][j] = '9'
  while children:
    child = children.pop()
    i, j = child
    adjs = ((i+1, j), (i-1, j), (i, j+1), (i, j-1))
    sum += 1
    for adj in adjs:
      i, j = adj
      if matrix[i][j] != '9':
        children.append(adj)
        positions.remove(adj)
        matrix[i][j] = '9'
  return sum

b = [0]*3
while positions:
  i, j = positions.pop()
  if matrix[i][j] != '9':
    r = flood((i, j))
    for i in range(3):
      if b[i] < r:
        b.insert(i, r)
        break

print(b[0] * b[1] * b[2])

# Part 1
# def smolest(x, args):
#   (i, j) = pos
#   adjs = ((i+1, j), (i-1, j), (i, j+1), (i, j-1))
#   for arg in args:
#     if x >= arg:
#       return False
#   return True;

# sum = 0
#     adjs = [matrix[(i, j+1)], matrix[(i, j-1)], matrix[(i-1, j)], matrix[(i+1, j)]]
#     if smolest(matrix[(i,j)], adjs):
#       sum += matrix[(i,j)]+1

File: day9/solution.cpp
#include <bits/stdc++.h>

using namespace std;
using std::cout;

array<pair<int, int>, 4> get_adjacents(pair<int, int> pos) {
  int i = pos.first;
  int j = pos.second;
  return array<pair<int, int>, 4>({
    make_pair(i+1, j),
    make_pair(i-1, j),
    make_pair(i, j+1),
    make_pair(i, j-1),
  });
}

long bfs(pair<int, int> pos, vector<vector<char>>& matrix) {
  long sum = 0;
  stack<pair<int, int>> children;
  children.push(pos);
  int i = pos.first;
  int j = pos.second;
  matrix[i][j] = '9';

  while (children.size() != 0) {
    auto child = children.top();
    children.pop();
    auto adjs = get_adjacents(child);
    i = child.first;
    j = child.second;
    ++sum;
    for (auto adj : adjs) {
      i = adj.first;
      j = adj.second;
      if (matrix[i][j] != '9') {
        children.push(adj);
        matrix[i][j] = '9';
      }
    }
  }
  return sum;
}

int main() {
  ifstream input("bb.input");
  string line;

  vector<vector<char>> matrix;
  while (getline(input, line)) {
    if (matrix.size() == 0) {
      matrix.push_back(vector<char>(line.length()+2, '9'));
    }
    istringstream iss(line);
    vector<char> n;
    n.push_back('9');
    for (auto c : line) {
      n.push_back(c);
    }
    n.push_back('9');
    matrix.push_back(n);
  }
  matrix.push_back(matrix[0]);

  vector<long> b{0, 0, 0};
  for (int i = 1; i < matrix.size()-1; ++i) {
    for (int j = 1; j < matrix[0].size()-1; ++j) {
      if (matrix[i][j] != '9') {
        long r = bfs(make_pair(i, j), matrix);
        for (int k = 0; k < 3; ++k) {
          if (b[k] < r) {
            b.insert(b.begin()+k, r);
            break;
          }
        }
      }
    }
  }

  printf("%ld\n", b[0] * b[1] * b[2]);
}